src/librustc_borrowck/borrowck/mir/dataflow/mod.rs

```diff
@@ -0,0 +1,1168 @@
+    ///
+    /// (For example, establishing the call arguments.)
+    ///
+    /// (Typically this should only modify `sets.on_entry`, since the
+    /// gen and kill sets should reflect the effects of *executing*
+    /// the start block itself.)
+    fn start_block_effect(&self, ctxt: &Self::Ctxt, sets: &mut BlockSets);
+
+    /// Mutates the block-sets (the flow sets for the given
+    /// basic block) according to the effects of evaluating statement.
+    ///
+    /// This is used, in particular, for building up the
+    /// "transfer-function" represnting the overall-effect of the
+    /// block, represented via GEN and KILL sets.
+    ///
+    /// The statement here is `idx_stmt.1`; `idx_stmt.0` is just
+    /// an identifying index: namely, the index of the statement
+    /// in the basic block.
+    fn statement_effect(&self,
+                        ctxt: &Self::Ctxt,
+                        sets: &mut BlockSets,
+                        bb: repr::BasicBlock,
+                        idx_stmt: usize);
+
+    /// Mutates the block-sets (the flow sets for the given
+    /// basic block) according to the effects of evaluating
+    /// the terminator.
+    ///
+    /// This is used, in particular, for building up the
+    /// "transfer-function" represnting the overall-effect of the
+    /// block, represented via GEN and KILL sets.
+    ///
+    /// The terminator here is `idx_term.1`; `idx_term.0` is just an
+    /// identifying index: namely, the number of statements in `bb`
+    /// itself.
+    ///
+    /// The effects applied here cannot depend on which branch the
+    /// terminator took.
+    fn terminator_effect(&self,
+                         ctxt: &Self::Ctxt,
+                         sets: &mut BlockSets,
+                         bb: repr::BasicBlock,
+                         idx_term: usize);
+
+    /// Mutates the block-sets according to the (flow-dependent)
+    /// effect of a successful return from a Call terminator.
+    ///
+    /// If basic-block BB_x ends with a call-instruction that, upon
+    /// successful return, flows to BB_y, then this method will be
+    /// called on the exit flow-state of BB_x in order to set up the
+    /// entry flow-state of BB_y.
+    ///
+    /// This is used, in particular, as a special case during the
+    /// "propagate" loop where all of the basic blocks are repeatedly
+    /// visited. Since the effects of a Call terminator are
+    /// flow-dependent, the current MIR cannot encode them via just
+    /// GEN and KILL sets attached to the block, and so instead we add
+    /// this extra machinery to represent the flow-dependent effect.
+    ///
+    /// Note: as a historical artifact, this currently takes as input
+    /// the *entire* packed collection of bitvectors in `in_out`.  We
+    /// might want to look into narrowing that to something more
+    /// specific, just to make the interface more self-documenting.
+    fn propagate_call_return(&self,
```

Comment:
Just to be clear: Are you suggesting that we try to encode the flow-dependent effects by attaching one or more new bitsets, distinct from the existing gen- and kill-set attached to each block, to represent them?  (Or, as you say, perhaps some other data-structure besides a bitset, but that could still do batching of the bitvector updates rather than an iterative loop over the index space)?

Its an interesting idea. I don't want to hold up this PR to investigate it though... at least not until profiles indicate that we're spending a lot of time in bitvector updating.